specs/hierarchical_consensus.md

```diff
@@ -0,0 +1,1082 @@
+// Returns the checkpoint template to be populated for the current
+// checkpoint window. The current checkpoint window is computed by 
+// (currEpoch / CheckPeriod) * CheckPeriod. This determines the checkpoint
+// that needs to be populated and signed for the current window.
+CheckpointTemplate()
+```
+
+As shown on the next figure, the checkpointing protocol has two distinct stages:
+- __Checkpoint Window__: In this window the `SCA` opens a checkpoint template and starts populating it with cross-messages arriving to the subnet (and that need to be propagated further inside the checkpoint); and with the aggregation of checkpoints committed from the current network's children. The checkpoint window for the checkpoint in epoch `n` starts at `n-CheckPeriod` and ends at epoch `n`.
+- __Signing Window__: The signing window is the time range reserved for validators of the subnet to populate the checkpoint template with the corresponding proof of the state of the checkpoint, to sign the checkpoint, and to submit it to the corresponding `SA` in the parent for committment. The signing window for the checkpoint in epoch `n` goes from `n` to `n+CheckPeriod`. Consequently, the signing window for the checkpoint of epoch `n-CheckPeriod` and the checkpoint window for the checkpoint of epoch `n` is run in parallel. The checkpoint template provided by the `SCA` has `CrossMsgs`, `PrevCheck`, and `Epoch` already populated, and validators only have to add the corresponding `Proof` and sign it.
+
+
+![](https://hackmd.io/_uploads/HkAodIuFc.png)
+
+### Checkpoints data structure
+Checkpoints are always identified though the [Content Identifier (CID)](https://github.com/multiformats/cid) of their `Data` (i.e. the payload of the checkpoint), and they can optionally include the corresponding signature from validators in the subnet chain (this can be the signature of an individual miner, a multi-signature, or a threshold signature, depending on the `SA` policy). The signature is never used for the computation of the CID of the checkpoint.
+
+> Checkpoints Data Structure
+```go
+
+// Checkpoints wrap the checkpoint data and
+// a field dedicated for the arbitrary signing policy
+// used by the subnet.
+type Checkpoint struct {
+    Data CheckpointData
+    Sig []byte
+}
+
+// Data included in checkpoints
+type CheckpointData struct {
+    // Source subnet of the checkpoint.
+    Source     address.SubnetID
+    // Proof of the state of the subnet. In the
+    // reference implementation this is the tipset
+    // for the epoch in the checkpoint, but this could
+    // be a ZKProof or any other proof chosen by the subnet.
+    Proof     []byte
+    // Epoch of the checkpoint. Determined by the
+    // subnet's CheckPeriod
+    Epoch     abi.ChainEpoch
+    // Cid of the previous checkpoint committed for the subnet.
+    PrevCheck cid.Cid
+    // Array with the aggregation of all the checkpoints committed
+    // from the subnet's children.
+    Children  []ChildCheck
```

Comment:
I'm slightly confused here. Is the checkpoint supposed to include an array of all the child checkpoints? What does the aggregation look like?